agents/postgres/pgstatmonitor/pgstatmonitor.go

```diff
@@ -0,0 +1,288 @@
+}
+
+const queryTag = "pmm-agent:pgstatmonitor"
+
+// New creates new PGStatMonitorQAN QAN service.
+func New(params *Params, l *logrus.Entry) (*PGStatMonitorQAN, error) {
+	sqlDB, err := sql.Open("postgres", params.DSN)
+	if err != nil {
+		return nil, err
+	}
+	sqlDB.SetMaxIdleConns(1)
+	sqlDB.SetMaxOpenConns(1)
+	sqlDB.SetConnMaxLifetime(0)
+
+	reformL := sqlmetrics.NewReform("postgres", params.AgentID, l.Tracef)
+	// TODO register reformL metrics https://jira.percona.com/browse/PMM-4087
+	q := reform.NewDB(sqlDB, postgresql.Dialect, reformL).WithTag(queryTag)
+
+	return newPgStatMonitorQAN(q, sqlDB, params.AgentID, params.DisableQueryExamples, l)
+}
+
+func newPgStatMonitorQAN(q *reform.Querier, dbCloser io.Closer, agentID string, disableQueryExamples bool, l *logrus.Entry) (*PGStatMonitorQAN, error) {
+	value, err := q.FindOneFrom(pgStatMonitorSettingsView, "name", "pg_stat_monitor.pgsm_normalized_query")
+	if err != nil {
+		return nil, err
+	}
+
+	return &PGStatMonitorQAN{
+		q:                    q,
+		pgsmNormalizedQuery:  value.(*pgStatMonitorSettings).Value == 1,
+		dbCloser:             dbCloser,
+		agentID:              agentID,
+		l:                    l,
+		changes:              make(chan agents.Change, 10),
+		monitorCache:         newStatMonitorCache(retainStatMonitor, l),
+		disableQueryExamples: disableQueryExamples,
+	}, nil
+}
+
+// Run extracts stats data and sends it to the channel until ctx is canceled.
+func (m *PGStatMonitorQAN) Run(ctx context.Context) {
+	defer func() {
+		m.dbCloser.Close() //nolint:errcheck
+		m.changes <- agents.Change{Status: inventorypb.AgentStatus_DONE}
+		close(m.changes)
+	}()
+
+	// add current stat monitor to cache so they are not send as new on first iteration with incorrect timestamps
+	var running bool
+	m.changes <- agents.Change{Status: inventorypb.AgentStatus_STARTING}
+	if current, _, err := m.monitorCache.getStatMonitorExtended(ctx, m.q, m.pgsmNormalizedQuery, m.disableQueryExamples); err == nil {
+		m.monitorCache.refresh(current)
+		m.l.Debugf("Got %d initial stat monitor.", len(current))
+		running = true
+		m.changes <- agents.Change{Status: inventorypb.AgentStatus_RUNNING}
+	} else {
+		m.l.Error(err)
+		m.changes <- agents.Change{Status: inventorypb.AgentStatus_WAITING}
+	}
+
+	// query pg_stat_monitor every minute at 00 seconds
+	start := time.Now()
+	wait := start.Truncate(queryStatMonitor).Add(queryStatMonitor).Sub(start)
+	m.l.Debugf("Scheduling next collection in %s at %s.", wait, start.Add(wait).Format("15:04:05"))
+	t := time.NewTimer(wait)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			m.changes <- agents.Change{Status: inventorypb.AgentStatus_STOPPING}
+			m.l.Infof("Context canceled.")
+			return
+
+		case <-t.C:
+			if !running {
+				m.changes <- agents.Change{Status: inventorypb.AgentStatus_STARTING}
+			}
+
+			lengthS := uint32(math.Round(wait.Seconds())) // round 59.9s/60.1s to 60s
+			buckets, err := m.getNewBuckets(ctx, start, lengthS)
```

Comment:
TBH, it does not look like we align buckets correctly. Why we start bucket at `:00` seconds when we have bucket start time, for example?